aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -0,0 +1,366 @@
+// . - - - - - - - - - - - - - .                    . - - - - - - - - - - - - - .
+//          |                   \                  /                      |
+//          |                     \              /                        |
+//          |                       . - - - - - - - - - - - - .           |
+//          |                       |  FunctionalStorageImpl  |           |
+//          |                       . - - - - - - - - - - - - .           |
+//          |                       |         Alias           |           |
+//          |                       . - - - - - - - - - - - - .           |
+//          |                       /     mutations to a or b             |
+//          |                     /       are queued onto Alias           |
+//          |                   /                                         |
+//         \/                 /                                           \/
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+// |        TensorImpl         |                             |             TensorImpl        |
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+// |   value   |   storage     |                             |    storage    |     Value     |
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+//          |                                                             |
+//          |                                                             |
+//          |                                                             |
+//          |   In this picture the two tensor views their own storages,  |
+//          |   have their own storages, but backends like functorch      |
+//         \/   are allowed to re-alias underneath the pass               \/
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+// |    underyling_storage     |                             |      underyling_storage       |
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+//
+// See Note [Functionalization: Alias Removal Part 2] for more details on the mutation replay logic.
+void FunctionalTensorWrapper::set_view_meta(const Tensor& other, at::functionalization::ViewMeta meta) {
+    TORCH_INTERNAL_ASSERT(at::functionalization::impl::isFunctionalTensor(other));
+    auto other_impl = at::functionalization::impl::unsafeGetFunctionalWrapper(other);
+    TORCH_INTERNAL_ASSERT(this != other_impl);
+    TORCH_INTERNAL_ASSERT(!is_aliased())
+    auto other_storage_impl = static_cast<functionalization::FunctionalStorageImpl*>(other.unsafeGetTensorImpl()->unsafe_storage().unsafeGetStorageImpl());
+    if (!other_storage_impl->is_aliased()) {
+        // The original tensor wasn't a view - turn it into a (no-op) view.
+        other_storage_impl->set_alias(other_impl->value());
```

Comment:
I'm still on a quest to eliminate `set_alias`, and this call is the big sticking point. Remember `set_view_meta` is called whenever we do a view operation on a functional tensor, and this can happen in parallel, and this does a mutation on the storage, so you *still* have the race condition here.

With the other changes you've done on this PR, though, we're closer to a good spot: the key is that we should *eagerly* populate the Alias structure on FunctionalStorageImpl. The downside to this is that we can't test if a storage is aliased or not, and so we have to always record updates for later sync in that case. But actually we can: storage is refcounted, so we can simply assume that there is no alias to storage if its refcount is 1. I'm looking to see if there are other consequences to doing this, but I can't see any others so far.

Reply:
(And just to distinguish this from the other race condition, which we're not going to fix, this race involves two concurrent "read" accesses: taking out views)

Reply:
i just completely forgot about this race condition when I did the re-write :(.

Checking the refcount of alias's storage sounds like it should work.

Reply:
This requires an dumb change to preserve mutations to inputs for functorch:

When exiting the functionalization pass, I call sync_() on every tensor input to ensure that updates to the alias have been propagated to the input tensors. By the time I'm exiting the functionalization pass though, any views created off the inputs have gone out of scope, so the use_count of the input FunctionalTensorWrapper will always be 1. The call to `sync_()` becomes a no-op, because the wrapper thinks that it isn't an alias.

I can just get around that by forcing a sync (unconditionally calling sync_update_operations and returning the updated alias), since we have no way to know at this point whether or not any temporary tensors inside of the program had alias'd the input.